onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,709 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/propagate_cast_ops.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/utils.h"
+#include <deque>
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+namespace onnxruntime {
+// The collection fp16_allow_ops, specifies for a given propagate_cast_ops level, a vector of node op_types that
+// the code is allowed to propage Cast operations cross. The user may specify a custom list of optypes using level 0.
+// The opcodes are split into multiple levels. Cast propagation is done based on the level. Level 2 op code
+// list includes Level 1 list also.
+static std::vector<std::unordered_set<std::string>> fp16_allow_ops = {
+    /* Level 0 */ {},
+    /* Level 1 */ {"Transpose", "Relu", "Reshape", "Split", "Tanh"},
+    /* Level 2 */ {"BiasGelu", "Dropout", "FastGelu", "Gather", "Gelu", "LayerNormalization", "Where"}};
+static std::vector<std::string> inserted_node_names;  // Names of the nodes inserted
+
+// Check whether the given opcode is fp16 allowed for the given level of optimization.
+static bool IsFP16Allow(const std::string& op_type, size_t level) {
+  bool fp16_allow = false;
+  for (size_t i = 0; i <= level && i < fp16_allow_ops.size() && !fp16_allow; ++i) {
+    fp16_allow = std::find(fp16_allow_ops[i].begin(), fp16_allow_ops[i].end(), op_type) != fp16_allow_ops[i].end();
+  }
+  return fp16_allow;
+}
+
+static bool IsCastTo(const Node* node, TensorProto_DataType data_type) {
+  if (node->OpType() == "Cast") {
+    const NodeAttributes& attributes = node->GetAttributes();
+    ORT_ENFORCE(attributes.find("to") != attributes.end());
+    return attributes.at("to").i() == static_cast<int64_t>(data_type);
+  }
+  return false;
+}
+
+static bool IsType(const NodeArg& node_arg, TensorProto_DataType data_type) {
+  return node_arg.TypeAsProto()->tensor_type().elem_type() == data_type;
+}
+
+// InsertCastNodes
+// Insert a new Cast node after each NodeArg in the require_cast vector. The cast node is FLOAT16 if is_fp16 is True
+// and FLOAT otherwise. This funtion fixes the graph edges in addition to inserting the cast nodes.
+static Status InsertCastNodes(Graph& graph,
+                              const std::unordered_set<NodeArg*>& require_cast,
```

Comment:
>NodeArg [](start = 55, length = 7)

Non blocking comment, Can this be const NodeArg*  ?

Reply:
Find the answer below.